### ISSUE TITLE
Swift 2 sink type changes behavior of calling start with trailing closure

### ISSUE THREAD

**Issue by sebreh** [ReactiveCocoa/ReactiveCocoa#2311]
In Swift 2 (and the *swift2* branch of this repo), the `SinkOf<T>` type is gone and instead a normal closure `T -> ()` is used. This is great but it introduces a potential risk due to the behaviour change when calling SignalProducer<T,E>.start() with a trailing closure. In Swift 1.x, the trailing closure would default to the `next` argument, and in order to provide a sink you would need to pass in a `SinkOf<Event<T>>`. In Swift 2 however, the trailing closure will match the new sink type `T -> ()`. From the call site, the call looks identical after migrating to the swift2 branch but the behaviour is different. This could potentially introduce issues in cases where the argument to start is ignored (`_`), as the type checker will not catch it and the closure will be called on *next/interrupted/error/completed* and not just on *next*.

Example:

Swift 1.x:

```swift
producer |> start { _ in
	// Block only called for next events, `_` is of type Bool
}
```

Swift 2:

```swift
producer.start { _ in
	// Block is called for next/completed/interrupted/error events, `_` is of type (Event<Bool, NoError>)
}
```

You could just argue to "deal with it", but I was considering to just re-add a `SinkOf<T>` type to wrap the sink closure and call `start(SinkOf { x in ... })`, like this:

```swift
public struct SinkOf<T> {
	let sink: T -> ()

	init(_ sink: T -> ()) {
		self.sink = sink
	}
}

SignalProducer<Bool, NoError>(value: true).start(SinkOf { e in
    // e is an Event<Bool, NoError>
})
```

It would also allow you to continue using `producer.start { next in ... }` without having to specify explicitly type of the `next` parameter to help the type system find the right `start` function.

What do you think? I'd be happy to submit a PR if you agree this needs fixing.

**Comment by neilpa** [comment]
Yes, this needs to be fixed. However, I'd like to find a way that doesn't require bringing back `SinkType`.

**Comment by sebreh** [comment]
I can see that. You could have a different function name, e.g. `startWithSink(sink: T-())`. Although that would make the API biased towards using the shorter and cleaner `start` with closures. At the same time, `startWithSink(sink: T-())` is not that much heavier than the `start(sink: SinkOf<T>)` we had before.

**Comment by neilpa** [comment]
Actually, the `start` and `observe` overloads should be removed. They were useful before because the `Box`ed case values made pattern matching less than pleasant. That's no longer a problem and they could be replaced with `startNext`, `startError` for the cases where you only care about a single case. Otherwise, work with the `Event` directly when you care about more than one case.

**Comment by NachoSoto** [comment]
I like that!

**Comment by sebreh** [comment]
Sounds like a good plan! I'll prepare a PR.

**Comment by neilpa** [comment]
Resolved by #2318
